src/pkcs15init/pkcs15-westcos.c

```diff
@@ -246,7 +246,8 @@ static int westcos_pkcs15init_generate_key(sc_profile_t *profile,
 	if (BN_set_word(bn, RSA_F4) != 1 ||
 	    EVP_PKEY_keygen_init(pctx) != 1 ||
 	    EVP_PKEY_CTX_set_rsa_keygen_bits(pctx, key_info->modulus_length) != 1 ||
-	    EVP_PKEY_CTX_set1_rsa_keygen_pubexp(pctx, bn) != 1 ||
+	    EVP_PKEY_CTX_set_rsa_keygen_pubexp(pctx, bn) != 1 ||
+	    (bn = NULL) == NULL || /* pctx will free bn */
```

Comment:
If "EVP_PKEY_CTX_set_rsa_keygen_pubexp(pctx, bn)"  fails , then bn is still set and needs to be freed. If  "EVP_PKEY_CTX_set_rsa_keygen_pubexp(pctx, bn)" passes -  man page says: "The pubexp pointer is used internally by this function so it should not be modified or freed after the call."

Reply:
> EVP_PKEY_CTX_ctrl() and its macros return a positive value for success and 0 or a negative value for failure. In particular a return value of -2 indicates the operation is not supported by the public key algorithm.

We should check for `< 1` to get it right in all cases.

But what I wanted to point out initially was that `(bn = NULL) == NULL` is always true so you will always get to the `r = SC_ERROR_UNKNOWN;` branch. I think you wanted to use `(bn = NULL) != NULL` or something similar.

Reply:
Neither `(bn = NULL) == NULL` (always true), nor `(bn = NULL) != NULL` (always false) make any sense.

Please explain what you're trying to accomplish there.

Reply:
You are right, fix committed.

Reply:
In reply to @mouse07410  
EVP_PKEY_CTX_set_rsa_keygen_pubexp  takes a pointer to a BN and adds it to the pctx. "The pubexp pointer is used internally by this function so it should not be modified or freed after the call."

The preexisting if statement expression is a series of logical or "||" operations. The operands are function calls and tests that set true or false. As long as segment is false it will try the next. If any return true it will get to the `r = SC_ERROR_UNKNOWN`.

The `EVP_PKEY_CTX_set_rsa_keygen_pubexp(pctx, bn) != 1` is the segment that stores the `bn` pointer value in the pctx.  The next segment, `(bn = NULL) != NULL` has an assigment and a test.  `(bn = NULL)` clears the pointer so it is not freed at the end of the function. The result of the `(bn = NULL) != NULL`  is "(always false)"  so the next segment of the if, `EVP_PKEY_keygen(pctx, &key) != 1` will be run.